Mage.Sets/src/mage/cards/i/IndominusRexAlpha.java

```diff
@@ -0,0 +1,183 @@
+
+    protected String getOtherChoicesAsString(){
+        StringBuilder sb = new StringBuilder();
+        CounterType[] choicesArray = copyableCounters.toArray(new CounterType[0]);
+
+        // Skip first counter name - flying
+        for (int i = 1; i < copyableCounters.size(); i++){
+            String separator = ", ";
+            if (i == copyableCounters.size()-2){
+                separator = ", and ";
+            } else if (i == copyableCounters.size()-1){
+                separator = "";
+            }
+            sb.append(choicesArray[i].getName()).append(separator);
+        }
+        return sb.toString();
+    }
+
+    private IndominusRexAlphaCountersEffect(final IndominusRexAlphaCountersEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public IndominusRexAlphaCountersEffect copy() {
+        return new IndominusRexAlphaCountersEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+
+        // EntersBattlefieldAbility is static, see AddCountersSourceEffect
+        Permanent permanent = game.getPermanentEntering(source.getSourceId());
+        if (permanent == null) {
+            return true;
+        }
+
+        TargetCard target = new TargetDiscard(0, Integer.MAX_VALUE, StaticFilters.FILTER_CARD_CREATURE, controller.getId());
+        controller.choose(outcome, controller.getHand(), target, source, game);
+        List<UUID> chosenTargets = target.getTargets();
+
+        // the basic event is the EntersBattlefieldEvent, so use already applied replacement effects from that event
+        List<UUID> appliedEffects = (ArrayList<UUID>) this.getValue("appliedEffects");
+
+        boolean addedCounters = false;
+
+        for (CounterType counterType : copyableCounters) {
+
+            Counter counter = counterType.createInstance();
+            Ability abilityToCopy = ((AbilityCounter)(counter)).getAbility();
+            Class<? extends Ability> abilityClass = abilityToCopy.getClass();
+
+            boolean addedCounterType = false;
+
+            for (UUID targetId : chosenTargets) {
+
+                Card card = game.getCard(targetId);
+                if (card == null){
+                    continue;
+                }
+
+                for (Ability ability : card.getAbilities()) {
+                    if (abilityClass.isInstance(ability)){
+                        permanent.addCounters(counter, source.getControllerId(), source, game, appliedEffects);
+                        addedCounterType = true;
+                        break;
+                    }
+                }
+
+                if (addedCounterType){
+                    addedCounters = true;
+                    break;
+                }
+
+            }
+
+        }
+        controller.discard(new CardsImpl(target.getTargets()), false, source, game);
```

Comment:
Wrong commands order. It discards first by rules text and add all counters at the end. So must use structure like:
- choose;
- check choices and find counters to add;
- discard;
- add counters.

Reply:
Point of order: should we be checking for the abilities before or after the cards have moved to the graveyard? Looking at the rules test, if for some reason a card gains hexproof only in the graveyard, it would be seen by indominus.

Reply:
Code order must be changed anyway (discard first then add counters).

> should we be checking for the abilities before or after the cards have moved to the graveyard

It’s interesting moment. Maybe a question for a judge chat. Found only one related card and rules from Recommission - it look after move:

![IMG_0276](https://github.com/magefree/mage/assets/8344157/3df8a237-00ed-40f0-a95a-142b84b9383b)
![IMG_0277](https://github.com/magefree/mage/assets/8344157/889ef990-d348-4ff0-920c-282bc4a43eb3)

BUT etb part can be related to blood moon (#4202) rules - etb creature affected by battlefield effects before enter. And it not related to another zones.

Reply:
I see--Where do we normally go to query for this ruling?

Reply:
[Judge Chat](https://chat.magicjudges.org/mtgrules/)

Reply:
For what it's worth, though, the phrasing is "if a card discarded this way *has* flying" i.e. present tense. So it should be checked on resolution.

Reply:
Ill try that chat, thanks!

> For what it's worth, though, the phrasing is "if a card discarded this way has flying" i.e. present tense. So it should be checked on resolution.

Can you rephrase this? I'm not 100% sure what you mean by this. resolution of the counters being added or the discarding?

Reply:
* IntelliJ IDEA regexp search for rules;
* [Scryfall search](https://scryfall.com/search?q=oracle%3A%22It+enters+with+a%22&unique=cards&as=grid&order=name) for cards with same rules;
* Official judge chat for rules questions: https://web.libera.chat/#magicjudges-rules

Reply:
> Ill try that chat, thanks!
> 
> > For what it's worth, though, the phrasing is "if a card discarded this way has flying" i.e. present tense. So it should be checked on resolution.
> 
> Can you rephrase this? I'm not 100% sure what you mean by this. resolution of the counters being added or the discarding?

Effects resolve in the order you read. So you discard, then you do the bit about putting counters. And since the rules so the cards *have* flying, you need to check the state of those cards right now - i.e. in the graveyard

Reply:
For info: if it will not work (do not see graveyard’s abilities) then must be called game.getstate.processActions after discard.

Reply:
Looks like it shouldn't matter because we only care about "printed abilities":
![image](https://github.com/magefree/mage/assets/17803304/44ab0320-6d69-4f51-b9c4-8688b58ed911)

When asked about a theoretical case, Blees said:
> since there's no actual card to compare it to, it's impossible to say. We can't really make rulings on cards tht don't exist

Reply:
It sounds like we care about what is in hand, since i'm not sure anything will ever give a card one of these 12 abilities while a card is in hand (other than its own printed abilities)

Reply:
For counters code example can look at code in `KathrilAspectWarper`

Reply:
> For info: if it will not work (do not see graveyard’s abilities) then must be called game.getstate.processActions after discard.

This is only the case if it gets abilites from cards in hand and then discards?

Reply:
It's for all use cases with move between zones in **single** effect (if you move something and want to check it in new zone -- it must apply all current effects first). Move to battlefield uses it inside. But moves to other zones -- don't.

Reply:
+1 proof for [alexander-novo's version](https://github.com/magefree/mage/pull/12119#discussion_r1583478628) (must check current status, not past). Wretched Bonemass:
> Wretched Bonemass has flying as long as an exiled card used to craft it has flying. The same is true for first strike, double strike, deathtouch, haste, hexproof, indestructible, lifelink, menace, protection, reach, trample, and vigilance.

> You may exile tokens you control as part of the materials required. However, because they aren't cards and won't stay in exile, any abilities that refer to what you "used to craft" the back faces won't refer to anything. (2023-11-10)

Reply:
Cool so it looks like:
1. Indominus rex checks abilities that the cards have in the hand, not in the graveyard
2. because of 1, calling `game.getstate.processActions` is not necessary--since we are checking the cards' states in the zone they started in, not after we moved them to the new zone

yes?

Reply:
Like that:
1. Call choose;
2. Collect selected cards to list;
3. Discard selected cards;
4. Call process actions;
5. Take selected cards and retain/filter it to cards from graveyard zone only;
6. Check abilities and add counters;

Reply:
Wouldnt that check for abilities in the graveyard? i thought we determined that it cares about abilities in hand.

Reply:
Yes, it must check cards in graveyard (discarded cards).

Reply:
> Looks like it shouldn't matter because we only care about "printed abilities": ![image](https://private-user-images.githubusercontent.com/17803304/326536349-44ab0320-6d69-4f51-b9c4-8688b58ed911.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0MjUzMDcsIm5iZiI6MTcxNDQyNTAwNywicGF0aCI6Ii8xNzgwMzMwNC8zMjY1MzYzNDktNDRhYjAzMjAtNmQ2OS00ZjUxLWI5YzQtODY4OGI1OGVkOTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDIxMTAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZmZWIwN2JlOGZiZDFiMGRhNWYyNDA3ZDlhNmNmYTk4M2I0N2JmZmJmNjdkODcxMDI3MDM4OGIxZGZhOTY4MTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RiJByywtKSyr6OHKEqGlRKDTg6pZW-ZZFFHISOz8Vbs)
> 
> When asked about a theoretical case, Blees said:
> 
> > since there's no actual card to compare it to, it's impossible to say. We can't really make rulings on cards tht don't exist

See the judge's comment here. It looks like this is not the case.

Reply:
> Looks like it shouldn't matter because we only care about "printed abilities":
> ![image](https://github.com/magefree/mage/assets/17803304/44ab0320-6d69-4f51-b9c4-8688b58ed911)
> 
> When asked about a theoretical case, Blees said:
> > since there's no actual card to compare it to, it's impossible to say. We can't really make rulings on cards tht don't exist

Well that's not theoretical [[Yixlid Jailer]]

Reply:
A new judge has given us a more concrete answer:
![image](https://github.com/magefree/mage/assets/17803304/932ef1f8-f171-4aec-8cea-15d80d543b3a)

It looks like we DO care about what abilities are on the cards in the graveyard. i will follow the procedure that @JayDi85 laid out earlier.